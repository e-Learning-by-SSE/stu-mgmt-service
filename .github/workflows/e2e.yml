name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
