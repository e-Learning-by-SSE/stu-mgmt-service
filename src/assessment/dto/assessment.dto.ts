import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
