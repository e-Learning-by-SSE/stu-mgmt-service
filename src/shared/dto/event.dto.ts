export class EventDto {
	event: string;
	timestamp: Date;
	payload?: Record<string, unknown>;
}
