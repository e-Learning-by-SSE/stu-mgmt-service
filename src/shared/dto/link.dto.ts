export class LinkDto {
	url: string;
	name: string;
}
