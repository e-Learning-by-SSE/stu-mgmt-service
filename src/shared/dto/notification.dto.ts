import { AssignmentId } from "../../course/entities/assignment.entity";
import { CourseId } from "../../course/entities/course.entity";
import { GroupId } from "../../course/entities/group.entity";
import { Event } from "../../course/events/events";
import { UserId } from "../entities/user.entity";

export class NotificationDto {
	event: Event;
	courseId: CourseId;
	assignmentId?: AssignmentId;
	groupId?: GroupId;
	userId?: UserId;
	payload?: Record<string, unknown>;
}
