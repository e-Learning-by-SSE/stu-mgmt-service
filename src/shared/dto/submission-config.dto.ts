import { IsNotEmpty } from "class-validator";

export class SubmissionConfigDto {
	@IsNotEmpty()
    tool: string;
    @IsNotEmpty()
	config: string;
}
