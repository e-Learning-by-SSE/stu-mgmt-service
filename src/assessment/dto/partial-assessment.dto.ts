import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
