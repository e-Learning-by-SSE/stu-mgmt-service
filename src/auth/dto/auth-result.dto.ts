import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
