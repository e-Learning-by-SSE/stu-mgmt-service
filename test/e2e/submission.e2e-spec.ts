import { SubmissionDto } from "../../src/submission/submission.dto";
import { COURSE_JAVA_1920 } from "../mocks/courses.mock";
import { SUBMISSION_USER_1_ASSIGNMENT_IN_PROGRESS_HOMEWORK_GROUP_1 } from "../mocks/submissions.mock";
import { TestSetup } from "../utils/e2e";

const course = COURSE_JAVA_1920;
const submission = SUBMISSION_USER_1_ASSIGNMENT_IN_PROGRESS_HOMEWORK_GROUP_1;

describe("Submission E2E", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	describe("POST-Requests of SubmissionController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(POST) /courses/:courseId/submissions/assignments/:assignmentId", () => {
			const route = (courseId, assignmentId) =>
				`/courses/${courseId}/submissions/assignments/${assignmentId}`;
			it("Adds a submission", () => {
				return setup
					.request()
					.post(route(course.id, submission.assignmentId))
					.send(submission)
					.expect(201)
					.expect(({ body }) => {
						const result = body as SubmissionDto;
						result.date = null;
						expect(result).toMatchSnapshot();
					});
			});
		});
	});

	describe("GET-Requests of SubmissionController (e2e)", () => {
		beforeAll(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(GET) /courses/:courseId/submissions", () => {
			const route = courseId => `/courses/${courseId}/submissions`;

			it("Retrieves all submissions", () => {
				return setup
					.request()
					.get(route(course.id))
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubmissionDto[];
						expect(result).toMatchSnapshot();
					});
			});
		});

		describe("(GET) /courses/:courseId/submissions/users/:userId", () => {
			const route = (courseId, userId) => `/courses/${courseId}/submissions/users/${userId}`;

			it("Retrieves all submissions of user", () => {
				return setup
					.request()
					.get(route(course.id, submission.userId))
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubmissionDto[];
						expect(result).toMatchSnapshot();
					});
			});
		});

		describe("(GET) /courses/:courseId/submissions/groups/:groupId/assignments/:assignmentId", () => {
			const route = (courseId, groupId, assignmentId) =>
				`/courses/${courseId}/submissions/groups/${groupId}/assignments/${assignmentId}`;

			it("Retrieves all submissions of a group for an assignment", () => {
				return setup
					.request()
					.get(route(course.id, submission.groupId, submission.assignmentId))
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubmissionDto[];
						expect(result).toMatchSnapshot();
					});
			});
		});

		describe("(GET) /courses/:courseId/submissions/users/:user/assignments/:assignments", () => {
			const route = (courseId, userId, assignmentId) =>
				`/courses/${courseId}/submissions/users/${userId}/assignments/${assignmentId}`;

			it("Retrieves the latest submission of a user (or their group) for an assignment", () => {
				return setup
					.request()
					.get(route(course.id, submission.userId, submission.assignmentId))
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubmissionDto;
						expect(result).toMatchSnapshot();
					});
			});
		});
	});

	describe("DELETE-Requests of SubmissionController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(DELETE) /courses/:courseId/submissions/assignments/:assignmentId", () => {
			const route = (courseId, assignmentId) =>
				`/courses/${courseId}/submissions/assignments/${assignmentId}`;
			it("Deletes all submissions of an assignment", () => {
				return setup
					.request()
					.delete(route(course.id, submission.assignmentId))
					.expect(200);
			});
		});
	});
});
