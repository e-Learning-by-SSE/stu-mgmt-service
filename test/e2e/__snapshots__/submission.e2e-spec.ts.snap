// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submission E2E GET-Requests of SubmissionController (e2e) (GET) /courses/:courseId/submissions Retrieves all submissions 1`] = `
Array [
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T08:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": null,
    "payload": null,
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T07:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": Array [
      Object {
        "name": "Link 1",
        "url": "http://example.url",
      },
      Object {
        "name": "Link 2",
        "url": "http://example.url",
      },
    ],
    "payload": Object {
      "error": "Something went wrong.",
    },
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
]
`;

exports[`Submission E2E GET-Requests of SubmissionController (e2e) (GET) /courses/:courseId/submissions/groups/:groupId/assignments/:assignmentId Retrieves all submissions of a group for an assignment 1`] = `
Array [
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T08:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": null,
    "payload": null,
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T07:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": Array [
      Object {
        "name": "Link 1",
        "url": "http://example.url",
      },
      Object {
        "name": "Link 2",
        "url": "http://example.url",
      },
    ],
    "payload": Object {
      "error": "Something went wrong.",
    },
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
]
`;

exports[`Submission E2E GET-Requests of SubmissionController (e2e) (GET) /courses/:courseId/submissions/users/:user/assignments/:assignments Retrieves the latest submission of a user (or their group) for an assignment 1`] = `
Object {
  "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
  "date": "2020-02-02T08:42:00.000Z",
  "displayName": "Max Mustermann",
  "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
  "links": null,
  "payload": null,
  "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
}
`;

exports[`Submission E2E GET-Requests of SubmissionController (e2e) (GET) /courses/:courseId/submissions/users/:userId Retrieves all submissions of user 1`] = `
Array [
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T08:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": null,
    "payload": null,
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
  Object {
    "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
    "date": "2020-02-02T07:42:00.000Z",
    "displayName": "Max Mustermann",
    "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
    "groupName": "Testgroup 1",
    "links": Array [
      Object {
        "name": "Link 1",
        "url": "http://example.url",
      },
      Object {
        "name": "Link 2",
        "url": "http://example.url",
      },
    ],
    "payload": Object {
      "error": "Something went wrong.",
    },
    "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
  },
]
`;

exports[`Submission E2E POST-Requests of SubmissionController (e2e) (POST) /courses/:courseId/submissions/assignments/:assignmentId Adds a submission 1`] = `
Object {
  "assignmentId": "b2f6c008-b9f7-477f-9e8b-ff34ce339077",
  "date": null,
  "displayName": "Max Mustermann",
  "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
  "links": Array [
    Object {
      "name": "Link 1",
      "url": "http://example.url",
    },
    Object {
      "name": "Link 2",
      "url": "http://example.url",
    },
  ],
  "payload": Object {
    "error": "Something went wrong.",
  },
  "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
}
`;
